### ISSUE TITLE
We're going to need a UI of some kind

### ISSUE THREAD

**Issue by Vectorrent** [0-5788719150923125/praxis#6]
https://reactflow.dev/examples/overview

Not entirely sure what I want or what we're going to need, but we need to somehow allow the user to "wire-together" different experts, scattered across the swarm. The user's node is going to need to have an awareness of which peers it is connected to, as well as the peers that his peers are connected to. Somehow, we need to display the cyclical connections in the computation graph.

**Comment by Vectorrent** [comment]
I added a basic diagram to the `web/` directory, and it will load at the local API automatically (default: http://localhost:2100). It doesn't do anything useful yet, though. I'm not sure how to integrate model data in a meaningful way.

**Comment by tearflake** [comment]
Does web interface need to be programmed in Python?

I can imagine doing server side Python tweaks to feed data to a Javascript UI interface for browsers. Does that seem acceptable?

**Comment by Vectorrent** [comment]
> Does web interface need to be programmed in Python?

No, but we will need to host the web server with something like Flask. This part already works, and can be accessed at http://localhost:2100 when you start the Praxis server.

I do have an idea of what I want it to look like now. Here's a simple example: 
![image](https://github.com/user-attachments/assets/d165827a-370c-4faf-9975-489457c7733a)

Basically, we have a controller that learns how to "route" through various experts/layers, at every forward pass. The end result is something like this: 
![image](https://github.com/user-attachments/assets/3160ea86-fc63-47fb-a825-e194a92114a3)

So, if you can imagine a unique number assigned to every expert in the swarm - I want to draw the active-route through each expert, per-token, at every forward pass. The end result will look a bit like "synapses firing in the brain", though slower - because they will only change every ~3 seconds or so. 

Anyway, I thought [React Flow](https://reactflow.dev/) looked like a decent solution for this, so I set up a demo page in the [web/](https://github.com/0-5788719150923125/praxis/tree/main/web) directory. I haven't done anything with this yet, mostly because I'm not familiar with React.

Also going to need some kind of chat interface, and configuration/settings management.

**Comment by Vectorrent** [comment]
> I can imagine doing server side Python tweaks to feed data to a Javascript UI interface for browsers. Does that seem acceptable?

Yes, this is acceptable and necessary. The Pytorch model will produce training metrics, configuration values, log messages, etc. We will need to pass all of that from the backend (Python) to the frontend (JavaScript/HTML/CSS).

**Comment by tearflake** [comment]
Can there be a hierarchy in that graph diagram you posted? And how many nodes are usually depicted?

**Comment by Vectorrent** [comment]
> Can there be a hierarchy in that graph diagram you posted?

Potentially? I'm sure that React Flow could represent a hierarchy of some sort, but I'm not sure how you're envisioning it would look. It would have to tie to the underlying data structures in some meaningful, interpretable way, IMO.

**Comment by tearflake** [comment]
Yes, it has to be interpretable.

I'm trying to fit the visualization  idea in this direction:
![socmedia_converted](https://github.com/user-attachments/assets/db0f2fae-17b6-48d0-86cb-ecd3b795e28b)

Does this inspire you of anything?

**Comment by tearflake** [comment]
Yes, it has to be interpretable.

I'm trying to fit the visualization  idea in this direction:
![socmedia_converted](https://github.com/user-attachments/assets/db0f2fae-17b6-48d0-86cb-ecd3b795e28b)

Does this inspire you of anything?

**Comment by Vectorrent** [comment]
Oooh, yeah, I can see the fractal hierarchy now. I could envision a world where you click on a node/expert, and "zoom in" to details about that particular layer, including its name, its hyperparameters, its owner, its IP address, its DHT address, your model's preference/confidence in that node's "expertise", etc. I could definitely see something like that being useful.

**Comment by tearflake** [comment]
I have a finished thing right now. Each node is represented by one HTML for circle contents, and one XML to connect it to child nodes. It is scalable, meaning it can host a gazillion nodes that can be constructed and fetched from server.

If you have idea on how to populate nodes, and if you can find it useful, I'm interested in helping out making this visualization.

We can also add other elements to the main window, in new frames, like chat between servers, or anything else, if you are interested.

**Comment by Vectorrent** [comment]
I also had an idea: what if we gave each node *depth*?
![image](https://github.com/user-attachments/assets/05725789-6e0c-4ec2-ab28-f85450f38509)
This is especially *interpretable* here, because our whole architecture is made possible by a [Mixture of Depths](https://arxiv.org/abs/2404.02258). Further, with depth as a visual element, we could easily represent physical features like magnitude, cosine similarity, semantic similarity, etc. It would be very cool.

> I have a finished thing right now. Each node is represented by one HTML for circle contents, and one XML to connect it to child nodes. It is scalable, meaning it can host a gazillion nodes that can be constructed and fetched from server.

I would love to see an example of your thing. I think for now, just populate it with dummy data. We're too early in development to really build a consistent interface yet. There are too many changing variables still.

**Comment by tearflake** [comment]
Here, download [this](https://drive.google.com/file/d/1ylW4yYwIWH843ivBvB03hVI6JFMhScjl/view?usp=sharing). The MIT licence for your purposes is included.

Unzip it to local server directory and open it in a browser. You will see meta instructions spanned through the tree, it should be all clear. Take look at the `./contents` directory to see how the data is fetched. Let me know if you have any questions about it.

Take your time, think of something cool and what else could be added, and let me know how can I help. I'm looking forward to collaborate.

**Comment by Vectorrent** [comment]
I have tentatively started work on a mobile app, which I'm calling "Axis". You can use Godot to open the project at [axis](https://github.com/0-5788719150923125/praxis/tree/main/axis).

While this is not a web app, per se - a Godot project can be exported to HTML. I'm not sure if we'll actually do that yet, or if we'll go in a different direction entirely. We may omit the web page entirely, in favor of a web API, exclusively (which already exists; it's how we interface with the Godot app).

**Comment by Vectorrent** [comment]
There is now a web UI, while the Godot project has largely fallen out of support. It would be nice to have both (or perhaps even unify them), but I just don't have the skill set for the Godot parts at this point. Anyway, there's no need to keep this issue open forever, when we have a functional and stable UI now.
